cmd/kubelet/app/options/options.go

```diff
@@ -281,6 +285,7 @@ func (f *KubeletFlags) AddFlags(mainfs *pflag.FlagSet) {
 	f.addOSFlags(fs)
 
 	fs.StringVar(&f.KubeletConfigFile, "config", f.KubeletConfigFile, "The Kubelet will load its initial configuration from this file. The path may be absolute or relative; relative paths start at the Kubelet's current working directory. Omit this flag to use the built-in default configuration values. Command-line flags override configuration from this file.")
+	fs.StringVar(&f.KubeletDropinConfigDirectory, "config-dir", f.KubeletDropinConfigDirectory, "Path to a directory to specify dropins, allows the user to optionally specify additional configs to overwrite what is provided by default and in the KubeletConfigFile flag")
```

Comment:
so this will not work if feature gates are set via the config file?

Reply:
as feature gates will be read after the config reading completed

Reply:
Further, feature gates from cli have the highest precedence. if they're specified, they'll override the config. So to enable this feature the kubelet will need to be passed the feature gate via cli (or not specify cli but set it in the kubelet config)

Reply:
this poses a weird interaction: we check the feature gate mid config rendering, so there's the possibility the kubelet config or cli specify the feature gate, but the drop in overrides to not. Or vice versa, the user hopes the drop in will turn on the feature gate.

Reply:
currently the flow goes

cli -> kubeletConfig -> kubelet.conf.d -> cli feature gates

as in, the base is the cli, then kubeletConfig overrides ... eventually we re-initialize cli feature gates. 

the weirdness comes in if cli sets the feature gate to off, and the kubelet config sets it to on. then, it would start off, turn on when kubelet config renders. kubelet.conf.d would be read because the feature gate is on, but then the cli comes back and overrides it. the eventual rendered kubelet configuration would have the feature gate off, but will have read the kubelet.conf.d

I think a way to fix this is to check the feature gate and kubeletDropInDir (
```
				if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletDropInConfig) && len(configDropinDir) > 0 {
					return fmt.Errorf("flag %s specified but featuregate %s missing, cannot start kubelet", kubeletFlags.KubeletDropinConfigDirectory, features.KubeletDropInConfig)
				}
```
[on line 235](https://github.com/kubernetes/kubernetes/pull/119390/files#diff-3a14a9838132c6668f4d100c26afa05b471b149c469edfece9b84fcd058c8a4bR235) after we reinitialize the map. Since we'll be failing anyway, doing the extra work of parsing the full configuration is alright I think. This will at least prevent us from getting in the weird state where kubelet's internal structure says feature gate is off, but it took the drop in dir into account as if it was.

> so this will not work if feature gates are set via the config file?

if we go the route above, then this will fix this case. I believe the kubelet.conf.d setting the feature gate could be present without the cli specifying it (as long as the cli didn't explicitly turn it off)

Reply:
I think the best way to solve it will be to remove the feature gate completely. @johnbelamaric what is the motivation for the feature gate when the feature is fully under the new configuration flag?

Reply:
So, what is the matrix for the UX with and without this fix?

| CLI FG | CLI  ConfigDir | File FG | File ConfigDir | Dir FG | Dir Dir| Result |
|----| ---- | --- |--- |--- |--- |--- |
| on | not set | not set | set | not set | not set | feature enabled, config dir set from file |
| on | not set | not set | set | not set | not set | feature enabled, config dir set from file |
| off | set | Not specified | Not specified | not specified | not specified | Error |
| on | set | Not specified | Not specified | not specified | set | Error |

etc.
...

Reply:
I think understanding what we *want* that to be should answer the question as to what to do here...

Reply:
> I think the best way to solve it will be to remove the feature gate completely. @johnbelamaric what is the motivation for the feature gate when the feature is fully under the new configuration flag?

Users need to explicitly acknowledge that the *know* the new flag they are using is an alpha feature. That's why it is feature gated in a feature like this (which is an admin feature).

Reply:
Can we just not allow setting this particular feature gate via any means other than CLI? If KubeletConfigDir is only a CLI flag, I would be OK with that.

Reply:
That is, just error out if someone tries to set this particular feature gate in the file or dir files. It achieves the goal of explicit acknowledgement that it is an alpha feature. And if they have to touch the flags anyway (as you said KubeletConfigDir is only flags), limiting the FG to the flags should not make it inaccessible in any given tooling.

Reply:
today my understanding is that feature gates set from CLI will override feature gates set from config file. If we have a single "special" FG, it will be very confusing.

Reply:
If to use this feature all FGs needs to be moved to command line it will also be not desireable and may break some scenarios. Let's say somebody wants to try the feature. They moved all FGs to CLI. And there is a DaemonSet that updates the FGs list in Config and restarts kubelet. This DaemonSet will stop working.

Reply:
Help text is insufficient. I may be convinced by that and some logged warnings or something, but that is not really forcing explicity, positive acknowledgement is better. In `kubectl` we use ENV vars for this. Basically you have to set a special ENV var or using the flag is treated as an error. So, like set KUBELET_ALPHA_CONFIGDIR or something.

Reply:
what if the cli defaulted the kubelet-config-dir "" to start? then a user would have to opt-in, and presumably would read the help text in deciding to do so

Reply:
> what if the cli defaulted the kubelet-config-dir "" to start? then a user would have to opt-in, and presumably would read the help text in deciding to do so

People copy and paste examples from other scripts and the internet. They only read the docs if there is a problem or they need to change it and don't know how. I think the env var could work, it's a pretty simple mechanism. Do you see an issue there?

Reply:
I am fine with the env, @SergeyKanzhelev ?

Reply:
env works for me as well. Nice alternative. I cannot come up with anything smarter.